src/core/cgroup.c

```diff
@@ -1828,10 +1879,139 @@ static void unit_remove_from_cgroup_realize_queue(Unit *u) {
+         * any new controllers at this layer. */
+        if (UNIT_ISSET(u->slice)) {
+                r = unit_realize_cgroup_now_enable(UNIT_DEREF(u->slice), state);
+                if (r < 0)
+                        return r;
+        }
+
+        target_mask = unit_get_target_mask(u);
+        enable_mask = unit_get_enable_mask(u);
+
+        /* We can only enable in this direction, don't try to disable anything.
+         */
+        if (!unit_has_mask_enables_realized(u, target_mask, enable_mask)) {
+                CGroupMask new_target_mask, new_enable_mask;
+
+                new_target_mask = u->cgroup_realized_mask | target_mask;
+                new_enable_mask = u->cgroup_enabled_mask | enable_mask;
+
+                r = unit_create_cgroup(u, new_target_mask, new_enable_mask);
+                if (r < 0)
+                        return r;
+
+                cgroup_context_apply(u, new_target_mask, state);
+                cgroup_xattr_apply(u);
+        }
+
+        return 0;
+}
+
+/* Controllers can only be disabled depth-first, from the leaves of the
+ * hierarchy upwards to the unit in question. */
+static int unit_realize_cgroup_now_disable(Unit *u, ManagerState state) {
+        Iterator i;
+        Unit *m;
+        void *v;
+
+        assert(u);
+
+        if (u->type != UNIT_SLICE)
+                return 0;
+
+        HASHMAP_FOREACH_KEY(v, m, u->dependencies[UNIT_BEFORE], i) {
+                CGroupMask target_mask, enable_mask;
+
+                if (UNIT_DEREF(m->slice) != u)
+                        continue;
+
+                /* The cgroup for this unit might not actually be fully
+                 * realised yet, in which case it isn't holding any controllers
+                 * open anyway. */
+                if (!m->cgroup_path)
+                        continue;
+
+                /* We must disable those below us first in order to release the
+                 * controller. */
+                if (m->type == UNIT_SLICE)
+                        unit_realize_cgroup_now_disable(m, state);
+
+                target_mask = unit_get_target_mask(m);
+                enable_mask = unit_get_enable_mask(m);
+
+                /* We can only disable in this direction, don't try to enable
+                 * anything. */
+                if (!unit_has_mask_disables_realized(m, target_mask, enable_mask)) {
+                        CGroupMask new_target_mask, new_enable_mask;
+                        int r;
+
+                        new_target_mask = m->cgroup_realized_mask & target_mask;
+                        new_enable_mask = m->cgroup_enabled_mask & enable_mask;
+
+                        r = unit_create_cgroup(m, new_target_mask, new_enable_mask);
+                        if (r < 0)
+                                return r;
+
+                        cgroup_context_apply(m, new_target_mask, state);
+                        cgroup_xattr_apply(m);
```

Comment:
these last three statements should probably be a function of its own, any then shared with the same three lines in unit_realize_cgroup_now().

Reply:
I'm probably just going to probably put them in unit_create_cgroup, we always seem to call these in tandem